src/StoreApi/Utilities/CartController.php

```diff
@@ -580,69 +594,117 @@ public function validate_cart_item( $cart_item ) {
 	/**
 	 * Validate all coupons in the cart and check for errors.
 	 *
-	 * @throws RouteException Exception if invalid data is detected.
+	 * @throws InvalidCartException Exception if invalid data is detected.
 	 */
 	public function validate_cart_coupons() {
 		$cart_coupons = $this->get_cart_coupons();
+		$errors       = [];
 
 		foreach ( $cart_coupons as $code ) {
 			$coupon = new \WC_Coupon( $code );
-			$this->validate_cart_coupon( $coupon );
+			try {
+				$this->validate_cart_coupon( $coupon );
+			} catch ( RouteException $error ) {
+				$errors[] = new \WP_Error( $error->getErrorCode(), $error->getMessage(), $error->getAdditionalData() );
+			}
+		}
+
+		if ( ! empty( $errors ) ) {
+
+			$error = new WP_Error();
+			foreach ( $errors as $wp_error ) {
+				$error->add(
+					$wp_error->get_error_code(),
+					$wp_error->get_error_message(),
+					$wp_error->get_error_data()
+				);
+			}
+
+			throw new InvalidCartException(
+				'woocommerce_coupons_error',
+				$error,
+				409
+			);
 		}
 	}
 
+	/**
+	 * Validate the cart and get a list of errors.
+	 *
+	 * @return WP_Error[] An array of WP_Errors describing the cart's error state.
+	 */
+	public function get_cart_errors() {
+		$errors = [];
```

Comment:
Rather than an array of `WP_Error`, why don't we use a single `WP_Error` instance?

Reply:
I wanted to be consistent with the other error getters like:
- `get_cart_item_errors`
- `get_cart_coupon_errors`

At this point, we don't use any of them 😅 

Should we consider deprecating them? Or we could refactor them all in case we want to _catch_ errors from a specific context.

Reply:
Also, I refactored the private `stock_exceptions_to_wp_errors` based on this idea.

Reply:
If they are not needed but let's finalise the new system first.

Reply:
yes if unused now.